tests/test_landings.py

```diff
@@ -290,3 +340,143 @@ def test_failed_landing_job_notification(
+        url=hg_server,
+        push_path=hg_server,
+        pull_path=hg_server,
+        access_group=SCM_LEVEL_3,
+        config_override={
+            "fix.fakefmt:command": "python /app/tests/fake_formatter.py",
+            "fix.fakefmt:pattern": "set:**.txt",
+            "fix.fakefmt:linerange": "--lines={first}:{last}",
+        },
+    )
+
+    hgrepo = HgRepo(hg_clone.strpath, config=repo.config_override)
+
+    upload_patch(1, patch=PATCH_FORMATTED_1)
+    upload_patch(2, patch=PATCH_FORMATTED_2)
+    job = LandingJob(
+        status=LandingJobStatus.IN_PROGRESS,
+        requester_email="test@example.com",
+        repository_name="mozilla-central",
+        revision_to_diff_id={"1": 1, "2": 2},
+        revision_order=["1", "2"],
+        attempts=1,
+    )
+
+    worker = LandingWorker(sleep_seconds=0.01)
+
+    assert worker.run_job(job, repo, hgrepo, treestatus, "landoapi.test.bucket")
+    assert (
+        job.status is LandingJobStatus.LANDED
+    ), "Successful landing did not set correct status"
+    assert job.formatted_replacements == (
+        [
+            "ed641c014313c3eab82d73c59f38e0ceccc30f22",
+            "0f740e7cc1f2e13bcc95e43de84931b23e0534b7",
+        ]
+    ), "Did not correctly save hashes of formatted revisions"
+
+    with hgrepo:
+        # Get repo root since `-R` does not change relative directory, so
+        # we would need to pass the absolute path to `test.txt`
+        repo_root = hgrepo.run_hg(["root"]).decode("utf-8").strip()
+
+        # Get the content of `test.txt`
+        rev2_content = hgrepo.run_hg(["cat", "--cwd", repo_root, "-r", "2", "test.txt"])
+        rev3_content = hgrepo.run_hg(["cat", "--cwd", repo_root, "-r", "3", "test.txt"])
+
+    # Assert `test.txt` was correctly formatted
+    assert rev2_content == TESTTXT_FORMATTED_1
+    assert rev3_content == TESTTXT_FORMATTED_2
+
+
+def test_format_patch_fail(
+    app, db, s3, mock_repo_config, hg_server, hg_clone, treestatusdouble, upload_patch
+):
+    """Tests automated formatting failures before landing."""
+    treestatus = treestatusdouble.get_treestatus_client()
+    treestatusdouble.open_tree("mozilla-central")
+    repo = Repo(
+        tree="mozilla-central",
+        access_group=SCM_LEVEL_3,
+        url=hg_server,
+        push_path=hg_server,
+        pull_path=hg_server,
+        config_override={
+            # Force failure by setting a formatter that returns exit code 1
+            "fix.fail:command": "exit 1",
+            "fix.fail:pattern": "set:**.txt",
+        },
+    )
+
+    hgrepo = HgRepo(hg_clone.strpath, config=repo.config_override)
+
+    upload_patch(1)
+    upload_patch(2)
+    job = LandingJob(
+        status=LandingJobStatus.IN_PROGRESS,
+        requester_email="test@example.com",
+        repository_name="mozilla-central",
+        revision_to_diff_id={"1": 1, "2": 2},
+        revision_order=["1", "2"],
+        attempts=1,
+    )
+
+    worker = LandingWorker(sleep_seconds=0.01)
```

Comment:
See `test_failed_landing_job_notification` for a workaround, in order to prevent a notification from triggering. `        "landoapi.landing_worker.notify_user_of_landing_failure` would need to be mocked. That or we would have to refactor `test_notifications`.